feat: ownership

enitrat opened the pull request:
Closes #34 


Time spent: 4 hours

hurrikaanig commented on src/ch03-01-what-is-ownership.md:
```diff
@@ -0,0 +1,327 @@
+Cairo implements an ownership system to ensure the safety and correctness of its compiled code.
+The ownership mechanism complements the linear type system, which enforces that objects are used exactly once.
+This helps prevent common operations that can produce runtime errors, such as illegal memory address
+references or multiple writes to the same memory address, and ensures the soundness of Cairo programs
```
Maybe we should explain soundness or provide a link that explains it

enitrat replied:
I feel like that would be too much for the introduction, + it's not directly related to the subject here 🤔

wraitii commented on src/ch03-01-what-is-ownership.md:
```diff
@@ -0,0 +1,327 @@
+### The `Drop` Trait
+
+You may have noticed that the `Point` type in the previous example also implements the `Drop` trait.
+By default, a value may not go out of scope unless it was previously moved. For example, the following code does not compile:
```
In particular, this reads a little weird to me. It would be more obvious to say that 'Upon going out of scope, the `_drop_` function is called on the type. Therefore, this function must exist for code to compile.

wraitii replied:
This I believe differs from Rust where it's fine to not have this Drop thing ? I'm actually not sure how that's handled

enitrat replied:
Drop is an entirely different concept in Rust as it just ensures that memory is freed.

fricoben reviewed (changes requested):
90% question and comments to be sure I understood (that's a good experiment tho to be sure a random dev understand). 

10% minor changes.

Gud PR 🔥

enitrat commented on src/ch03-00-understanding-ownership.md:
```diff
@@ -0,0 +1,7 @@
+# Understanding Cairo's linear type system
+
+Cairo is a language built around a linear type system that allows us to
+statically ensure that in every Cairo program, a value is used exactly once.
```
No, here it's just for the linear type system. It has nothing to do with immutable memory.
It just makes sure that each instantiated variable is consumed (moved or dropped)

enitrat replied:
added: "This linear type system helps preventing runtime errors by ensuring that operations that could cause such errors, such as writing twice to a memory cell, are detected at compile time."

enitrat commented on src/ch03-00-understanding-ownership.md:
```diff
@@ -0,0 +1,7 @@
+Cairo is a language built around a linear type system that allows us to
+statically ensure that in every Cairo program, a value is used exactly once.
+This is achieved by implementing an ownership system
+and forbidding copying and dropping values by default. In this chapter, we’ll
```
Why do you think this? This is actually at the core of this ownership/linear type system

wraitii replied:
It's not obvious what that means, particularly 'dropping values', unless you've already read the chapter. Seems odd as an intro to me.

It might make more sense if you're coming from rust but I'm not sure that this should be assumed.

But then again I'm not sure that introducing this as a 'linear type system' helps either because I have no clue what the means and reading wikipedia doesn't help, so maybe this comment is irrelevant as a whole.

wraitii replied:
(And by 'this comment' I mean my comments on this section, though I am not certain that this introduction works really well if one needs to be deep into type theory to actually understand it.

enitrat replied:
> though I am not certain that this introduction works really well if one needs to be deep into type theory to actually understand it.

Here's how I see it: Readers who are familiar with the concept will be able to grasp in a second what this chapter is about. Readers who are unfamiliar will read the rest to understand the terms introduced here.

Providing an explanation about what a linear type system is is not in the scope of the book

enitrat commented on src/ch03-01-what-is-ownership.md:
```diff
@@ -0,0 +1,327 @@
+
+### Variable Scope
+
+Now that we’re past basic Cairo syntax, we won’t include all the `fn main() {`
+code in examples, so if you’re following along, make sure to put the following
+examples inside a `main` function manually. As a result, our examples will be a
+bit more concise, letting us focus on the actual details rather than
+boilerplate code.
```
Not sure as it doesn't belong in the section above either.

enitrat replied:
ok

Th0rgal commented:
Nice some reading before bed

enitrat commented on src/ch03-01-what-is-ownership.md:
```diff
@@ -0,0 +1,327 @@
+The ownership mechanism complements the linear type system, which enforces that objects are used exactly once.
+This helps prevent common operations that can produce runtime errors, such as illegal memory address
+references or multiple writes to the same memory address, and ensures the soundness of Cairo programs
+by checking at compile time that all the dictionaries are squashed.
```
Does it become clearer after reading the rest? I don't really want to be very descriptive in the introduction 🤔 
It's just an "abstract" concept behind the design of the language but it justifies this whole page.

enitrat commented on src/ch03-01-what-is-ownership.md:
````diff
@@ -0,0 +1,327 @@
+let arr = ArrayTrait::<u128>::new();
+```
+
+You can append values to an `Array` using the `append` method:
````
well `append` is self-explanatory 😆  its a synonym to `add`

enitrat commented on src/ch03-01-what-is-ownership.md:
```diff
@@ -0,0 +1,327 @@
+you would attempt to try to write to the same memory cell twice, which is not allowed in Cairo.
+To prevent this, the ownership of the `arr` variable moves from the `main` function to the `foo` function when it is first called. When trying to call `foo` with `arr` as a parameter again, the
+ownership of `arr` was already moved to the first call, so the ownership system prevents us from
+using the same instance of `arr` in `foo`.
```
yes exactly
would it make more sense if I had two different function names for the sake of the example?

enitrat commented on src/ch03-01-what-is-ownership.md:
```diff
@@ -0,0 +1,327 @@
+
+### The `Copy` Trait
+
+While Arrays and Dictionaries can't be copied, custom types that don't contain either of them can be.
```
types that don't contain an Array or a Dict, unlike this one for ex
```
struct MyType{
  arr:Array<u128>
}
```

enitrat commented on src/ch03-01-what-is-ownership.md:
````diff
@@ -0,0 +1,327 @@
+You can implement the `Copy` trait on your type by adding the `#[derive(Copy)]` annotation to your type definition. However, Cairo won't allow a type to be annotated with Copy if the type itself or any of its components don't implement the Copy trait.
+
+```rust
+#[derive(Copy, Drop)]
````
You can still pass values by reference, this is covered in the next page. But yes almost every custom type will derive these two trait implementations

enitrat commented on src/ch03-01-what-is-ownership.md:
````diff
@@ -0,0 +1,327 @@
+fn main() {
+    A {}; // error: Value not dropped.
+}
+```
````
because `A` doesn't derive the `Drop` implementation, so the compiler doesn't know how to drop it

Th0rgal commented:
Okay I'm hyped for the snapshots now

enitrat commented on src/ch03-01-what-is-ownership.md:
```diff
@@ -0,0 +1,327 @@
+
+When a variable goes out of scope, its value is dropped, unless ownership of the value has been moved to another variable.
+
+While this works, taking ownership and then returning ownership with every
+function is a bit tedious. What if we want to let a function use a value but
+not take ownership? It’s quite annoying that anything we pass in also needs to
+be passed back if we want to use it again, in addition to any data resulting
+from the body of the function that we might want to return as well.
```
This is actually the goal of this sentence - demonstrate how annoying and confusing it is to proceed this way to demonstrate how powerful borrowing with `ref` and `@` are.

enitrat commented:
@fricoben  - Comments addressed.
@wraitii - I reworked the `Drop` part, and introduced the part about `Destructors` that I forgot. I felt that it make more sense to introduce them in this order. It should be way more clear now, wdyt?

wraitii commented on src/ch03-01-what-is-ownership.md:
```diff
@@ -0,0 +1,383 @@
+
+Now, when `A` goes out of scope, its dictionary will be automatically `squashed`, and the program will compile.
+
+#### Copy Array data with Clone
```
This is a `####` instead of a  `###` which seems a little odd, as it implies it's under 'Destruct'

wraitii commented:
> @wraitii - I reworked the `Drop` part, and introduced the part about `Destructors` that I forgot. I felt that it make more sense to introduce them in this order. It should be way more clear now, wdyt?

Agreed, I think this is much more understandable now.
I'm just wondering what the `drop` function actually does? Can one manually implement it? Is it actually any different from `Destruct` then?

enitrat commented:
> I'm just wondering what the drop function actually does? Can one manually implement it? Is it actually any different from Destruct then?

As you said it's a "no-op" that just transfers ownership in the void. It compiles to the `drop` libfunc in Sierra but doesn't generate any CASM code.

fricoben reviewed (approved):
letzzz go
